apis/v1alpha2/shared_types.go

```diff
@@ -0,0 +1,356 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha2
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// ParentRef identifies an API object (usually a Gateway) that can be considered
+// a parent of this resource (usually a route). The only kind of parent resource
+// with "Core" support is Gateway. This API may be extended in the future to
+// support additional kinds of parent resources, such as HTTPRoute.
+type ParentRef struct {
+	// Group is the group of the referent.
+	//
+	// Support: Core
+	//
+	// +kubebuilder:default=gateway.networking.k8s.io
+	// +optional
+	Group *Group `json:"group,omitempty"`
+
+	// Kind is kind of the referent.
+	//
+	// Support: Core (Gateway)
+	// Support: Custom (Other Resources)
+	//
+	// +kubebuilder:default=Gateway
+	// +optional
+	Kind *Kind `json:"kind,omitempty"`
+
+	// Namespace is the namespace of the referent. When unspecified (or empty
+	// string), this will either be:
+	//
+	// * local namespace of the route when scope is set to Namespace.
+	// * no namespace when scope is set to Cluster.
+	//
+	// Support: Core
+	//
+	// +optional
+	Namespace *Namespace `json:"namespace,omitempty"`
+
+	// Scope represents if this refers to a cluster or namespace scoped
+	// resource. This may be set to "Cluster" or "Namespace".
+	//
+	// Support: Core (Namespace)
+	// Support: Custom (Cluster)
+	//
+	// +kubebuilder:validation:Enum=Cluster;Namespace
+	// +kubebuilder:default=Namespace
+	// +optional
+	Scope *string `json:"scope,omitempty"`
+
+	// Name is the name of the referent.
+	//
+	// Support: Core
+	//
+	// +kubebuilder:validation:MinLength=1
+	// +kubebuilder:validation:MaxLength=253
+	Name string `json:"name"`
+
+	// SectionName is the name of a section within the target resource. In the
```

Comment:
This is a good question. Although I'd like to simplify the API, I think this particular field is more helpful than it might seem on the surface. For HTTP and TLS routes, the hostnames field provide a means of connecting Routes to the subset of Listeners that support at least one of their hostnames. Of course that does require each Listener and Route to specify meaningful hostnames. Although less precise, this could potentially pass for initial usage.

Unfortunately this doesn't work for L4 routing where we don't have a hostname concept. In that case, `sectionName` is our only mechanism to attach a Route to a specific listener. This becomes especially important with the lack of L4 matching capabilities - we're really expecting a 1:1 Listener - Route relationship at this level.

Even for L7, while hostnames could act as an alternative to `sectionName`, they don't provide the same level of control that `sectionName` does. I think matches that relied purely on hostnames could become difficult to manage.